### DISCUSSION TITLE
Decoder wakeup formula

### DISCUSSION THREAD

**Discussion by jdlvfr** [MusicPlayerDaemon/MPD#1803] [Q&A] [answered]
Hi all,
player/Thread.cxx line 1052 is :

/* this formula should prevent that the decoder gets woken up
	   with each chunk; it is more efficient to make it decode a
	   larger block at a time */
	if (!dc.IsIdle() && dc.pipe->GetSize() <= decoder_wakeup_threshold) {
		if (!decoder_woken) {

If I understand correctly this formula doesn't work very well when the decoder is at next song as current song and next song shares the same buffer. Something like this should works better:

	/* this formula should prevent that the decoder gets woken up
	   with each chunk; it is more efficient to make it decode a
	   larger block at a time */
	int pipe_size = dc.pipe->GetSize();
	if (IsDecoderAtNextSong()) pipe_size += pipe->GetSize();
	
	if (!dc.IsIdle() && pipe_size <= decoder_wakeup_threshold) {	
   		if (!decoder_woken) {
Thanks

**Comment by MaxKellermann** [comment]
I don't get it. Why do you believe it doesn't work well? Why does your formula work better?

**Reply by jdlvfr** [reply]
Sorry should be more clear with the full code.

When the decoder switches to next song the buffer is already full of current song decoded chunks, the next song decoder pipe doesn't have enough room to expand to the threshold , decoder_woken variable is never reset to false and the decoder is only woken up once.

Threshold should be tested against the buffer fill or the sum of the two decoder pipes, pipe->GetSize() and dc.pipe->GetSize()

Test shows a better operation but I may have missed some things... 

/* this formula should prevent that the decoder gets woken up
	   with each chunk; it is more efficient to make it decode a
	   larger block at a time */
	if (!dc.IsIdle() && dc.pipe->GetSize() <= decoder_wakeup_threshold) {
		if (!decoder_woken) {
			decoder_woken = true;
			dc.Signal();
		}
	} else
		decoder_woken = false;

**Reply by MaxKellermann** [reply]
And what's the problem with that?

**Comment by jdlvfr** [comment]
Maybe it can help a very slow machine, but you are right it certainly won't hurt too much...

**Reply by MaxKellermann** [reply, marked as answer]
If you didn't observe a problem, then there is none. Your post sounded like you were trying to solve a problem, but apparently you were just reading the MPD source code trying to figure out how it works. And you did figure it out. Kudos for that!

No, it won't help with "a very slow machine", unless you mean a computer that's 25 years old or so? Practically all existing hardware is massively overpowered for MPD. MPD is 20 years old, and even back then, all computers were overpowered. It's just about decoding some music!

**Reply by jdlvfr** [reply]
Ok, thank you!
